### ISSUE TITLE
Conan recipe custom step

### ISSUE THREAD

**Issue by markovandooren** [conan-io/conan#7747]
@jgsogo As requested, this is a description of my Conan/Visual Studio integration.
I tried to be as detailed as possible such that it is also clear what I am
trying to achieve and why. Perhaps there is a better way to do all of this.

Standalone Conan
================
In the standalone Conan scenario, as used on the build server, 
the build() method does the following:

    def build(self):
        ...
        MSBuild(...)
        
        self.index_sources()
        ...

    def index_sources(self):
        # Use self.deps_cpp_info and self.deps_user_info
        # to create the source indexing stream of the PDB.
        #
        # For header files, or static libraries, we use
        # information from the dependencies to create the
        # source indexing mapping. Each package inserts the
        # the type of version control, the repository URL,
        # and the revision into deps_user_info.
        ...

Visual Studio Standard Conan Workflow
=====================================
A. DLLs

In the local development workflow, everything works fine for DLLs and header files. 
For DLLs, the PDB that is next to the DLL is used. The header files are part of the packages.
The propertysheets generated by Conan ensure that Visual Studio can find everything it needs.

B. LIBs

For static libraries, however, this does not work. Suppose program P uses a static
library of package A. That static library contains the information that it is compiled
from C:\whatever\the\path\was\on\the\build\server\a.cpp. Even if the developer would
have a.cpp locally, it is not in the same directory as that of the build server, and 
possibly not the same revision.

When our Conan package does the source indexing for the static library, it inserts the 
mapping for C:\whatever\the\path\was\on\the\build\server\a.cpp. Visual Studio, however,
does not create source indexing. That is done by external tools. Therefore its build
output does not contain the source indexing for a.cpp. If you want to debug and
step into the code of a.cpp, Visual Studio will ask you to locate the source manually.
So even though the Conan file of package A has inserted the source indexing into the PDB,
it is not being used.

Visual Studio Extra Conan Integration
=====================================
To solve this problem, I change our custom generator and our shared Conan code.

The generator creates an additional propertysheet that triggers the source indexing
for a binary that was built.This propertysheet is included by the generated C++ Conan propertysheet.

    <?xml version="1.0" encoding="utf-8"?>
    <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
        <Target Name="IndexSources" AfterTargets="Build">
            <Message Importance="High" Text="Indexing sources for $(TargetName)."/>
            <Exec Command="cd &quot;D:\dev\workspace\platform&quot; &amp; set SOURCE_INDEXING_TARGET=$(OutDir)\$(TargetName)&amp; conan build ."/>
        </Target>
    </Project>

We somehow have to tell Conan do call the index_source() method described above because
it has all the knowledge about the dependencies to know where the PDB is on the local system.

Problem
=======
Since we cannot just call the index_sources() method, we have to abuse nother method.
I thought that the least inappropriate place would be the build() method. If the build()
method detects that it is not in the local cache, we let it skip the MSBuild step, and
just do the source indexing.

In addition, we have to be able to pass information about what binary to source index.
Some binaries, such as unit tests, are not part of package P, to the Conan file does
not know about them. Also, because builds run in parallel, we cannot just index all of them
because then it can start indexing a PDB that Visual is still creating and that will
give an error. 

We now use an environment variable SOURCE_INDEXING_TARGET to pass this information.
If we are not building in the local cache, we index only that target. Otherwise, we
indexing everything that is part of the package. This is all quite hacky, and I have broken
the regular build server behavior at times when developing this.

Context
=======
The context for this particular custom method would be

1. The context as it is at the start of the build() method. It needs the deps_cpp_info,
   deps_user_info, and the information about the current package.
2. An externally provided name that indicates what must be source indexed.

At first sight it seems that the context at the start of build() is sufficient for most 
if not all customizations. So if I could run e.g.

    conan --context=build --target=index_sources "my source indexing target"

The corresponding method in the Conan file would then be:

    def index_sources(self, source_indexing_target):

That would give a lot of possibilities for integrating parts of the proces into
the local development toolchain. Some custom steps might need multiple arguments.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @markovandooren 

Are you aware of the ``toolchain()`` new method? This method is called for every ``conan install`` without needing to build or create a package. It can create custom files and build systems integrations. We are moving our own built-in integrations to this approach, it is likely that it will be the proposed one for Conan 2.0. 

Please have a look at: https://docs.conan.io/en/latest/creating_packages/toolchains.html

As an example of the way we are thinking with toolchains, this is a very early stage proof of concept, in which a ``toolchain()`` can do customized toolchain and customize generators to hand custom Visual Studio configurations like ``ReleaseDLL`` without needing to change the Conan settings: https://github.com/conan-io/conan/pull/7754

**Comment by markovandooren** [comment]
Hi @memsharded, I looked at the toolchain documentation, but it is not clear to me how I could use that for the use case in this issue. Maybe I am missing something. As far as I understand from the documentation, they generate files and are executed when you run 'conan install'. The hook that I run when building in Visual Studio needs to be executed for each project in the solution and should not install anything. 

Exporting additional files by using a toolchain does not help because there is no tool that could interpret these files. The source indexing Python code in conanfile.py is custom written to handle Conan dependencies. Of course I could move the code for source indexing out of conanfile.py and export all dependency information with an extra generator, and then parse it all again in the external source indexing script. But that seems to defeat the purpose of having a Python package definition.

More generally, a local developer workflow will typically do a few things that conanfile.py does and let Conan handle the rest. That means that it needs to be able to run one or more (sub)steps from the entire packaging workflow. The ability to call a custom method makes it possible to create these steps as separate methods that can be called both by the standard Conan workflow or from the outside by whatever tools the developer is using.

**Comment by memsharded** [comment]
Sorry this was not followed up back then.
Now this ticket is outdated, many of the above has changed in Conan 2, almost 2 years old now:

- new recipe methods, like ``generate()`` instead of ``toolchain()``
- new ``finalize()`` method for a final local installation/customization step.
- Local ``deployers`` and ``deploy()`` method.

Closing this ticket as outdated.
If there are any questions or issues, please open a new ticket against Conan 2, thanks for your feedback.
